### PR TITLE
feat!: allows selective loading of extensions using optional dependencies

### PR DESCRIPTION
**Pull request by SecondSkoll**
BREAKING CHANGE: The [full] optional dependency needs to be loaded to support the functionality that was previously bundled into the extension. This is to support a minimal install.

Adds a `[full]` optional dependency in `pyproject.toml` that is the equivalent of the previous version.
Only sets up extensions that are installed and detected, so other subsets of deps can be supported if necessary.
Changes `display_contributors` default to `False` to support the minimal install option.
Adds a test for minimal install.

### CODE REVIEW

**Comment by SecondSkoll** [comment]
The CLA check does not play nicely with GitHub's private email settings. Had to force change some commits, but should be all good now.

File: pyproject.toml [hunk 1]
```diff
@@ -27,25 +17,43 @@ requires-python = ">=3.8"
 [project.scripts]
 canonical-sphinx-hello = "canonical_sphinx:hello"
 
+# Moving extended version to optional install, but first iteration will need to use previous full install
+# Once deployed on pypi, can change canonical-sphinx==0.1.0 to canonical-sphinx[full]
 [project.optional-dependencies]
+full = [
```
[the remark this answers is not in the document]
Comment (SecondSkoll) [reply pyproject.toml:23]: It does not. The way `pyproject.toml` works is that this provides the ability to load additional optional dependencies.

The main dependencies are listed on line 4:

```
dependencies = [
    "sphinx>=7.1.2,<8",
    "furo",
    "myst-parser",
]
```
This is the minimal installation - as Sphinx and Furo are needed for the visual theme - and pretty much everyone seems to use the myst parser - so it felt silly to remove it.

Installing `canonical-sphinx[full]` then grabs the additional dependencies of:

```
full = [
    "linkify-it-py",
    "canonical-sphinx-extensions",
    "sphinx-copybutton",
    "sphinx-design",
    "sphinx-notfound-page",
    "sphinx-reredirects",
    "sphinx-tabs",
    "sphinxcontrib-jquery",
    "sphinxext-opengraph",
    "pyspelling",
]
```

The other options are specific iterations meant for some of the tests that were already written. As they were previously reliant on the full list of dependencies (i.e. all the packages listed in full were previously in the main dependency list), I added canonical==0.1.0 as a dep so they would get all the right packages. I had issues using `.[full]` and `canonical-sphinx[full]` for some reason.

File: pyproject.toml [hunk 2]
```diff
@@ -27,25 +17,43 @@ requires-python = ">=3.8"
...
     "codespell[toml]==2.3.0",
     "ruff==0.5.7",
     "yamllint==1.35.1"
 ]
 types = [
+    "canonical-sphinx==0.1.0",
     "mypy[reports]==1.11.2",
     "pyright==1.1.378",
 ]
 docs = [
```
[the remark this answers is not in the document]
Comment (SecondSkoll) [reply pyproject.toml:55]: The `[types]` optional dependencies are used by the `[testenv:lint-{mypy,pyright}]` step in tox.

The main package is the minimal install - though I think we could add an empty `[minimal]` install ref to potentially avoid confusion - but I don't think it would be that practical.

File: tox.ini [hunk 3]
```diff
@@ -115,7 +115,7 @@ description = Run pre-commit on staged files or arbitrary pre-commit commands (t
 commands = pre-commit {posargs:run}
 
 [docs]  # Sphinx documentation configuration
-extras = docs
+extras = full, docs
```
[the remark this answers is not in the document]
Comment (SecondSkoll) [reply tox.ini:118]: They are used together in the `[testenv:lint-docs]` test in tox - but otherwise the `[docs]` dependencies are unnecessary.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (SecondSkoll) [reply pyproject.toml:55]: You're right. Though as far as I can tell the `[docs]` dependency set was never actually needed for the RTD build. It could easily be bundled into `[full]`. We'd need to drop Python 3.8 support to bundle in the more recent versions of autobuild, but that's something we need to address anyway as 3.8 goes EOL.

File: canonical_sphinx/config.py [hunk 4]
```diff
@@ -160,7 +175,7 @@ def config_inited(_app: Sphinx, config: Any) -> None:  # noqa: ANN401
         ("github_issues", "enabled"),
         ("discourse", "https://discourse.ubuntu.com"),
         ("sequential_nav", "none"),
-        ("display_contributors", True),
+        ("display_contributors", False),
```
Comment (tigarmo) [inline canonical_sphinx/config.py:178]: what is this change?

File: docs/index.rst [hunk 5]
```diff
@@ -12,26 +12,16 @@ StarCraft
...
-   :reverse:
-
-   .. grid-item-card:: :ref:`Reference <reference>`
-
-      **Technical information** about Starcraft
-
-   .. grid-item-card:: :ref:`Explanation <explanation>`
-
-      **Discussion and clarification** of key topics
+.. list-table::
```
Comment (tigarmo) [inline docs/index.rst:15]: how come this was changed?

File: pyproject.toml [hunk 6]
```diff
@@ -27,25 +17,43 @@ requires-python = ">=3.8"
 [project.scripts]
 canonical-sphinx-hello = "canonical_sphinx:hello"
 
+# Moving extended version to optional install, but first iteration will need to use previous full install
+# Once deployed on pypi, can change canonical-sphinx==0.1.0 to canonical-sphinx[full]
```
Comment (tigarmo) [inline pyproject.toml:21]: is this comment still valid?

**Review by tigarmo — COMMENTED** [review]
nice thanks! Next cycle we should prioritize getting some tests going on this repo (there's currently only a big integrated test)

File: tox.ini (hunk 3 above)
Comment (tigarmo) [inline tox.ini:118]: With the inclusion of `"canonical-sphinx[full]"` in `docs` is this change still necessary?

File: docs/index.rst (hunk 5 above)
Comment (SecondSkoll) [reply docs/index.rst:15]: Grid items are provided by the sphinx design extension which is not in the mininal install, so the build will fail if these roles are included in the test set.

File: pyproject.toml (hunk 6 above)
Comment (SecondSkoll) [reply pyproject.toml:21]: No, I did end up fixing it. I'll remove the comment in the morning.

File: tox.ini (hunk 3 above)
Comment (SecondSkoll) [reply tox.ini:118]: Good catch, this change is now unnecessary

File: canonical_sphinx/config.py (hunk 4 above)
Comment (tigarmo) [reply canonical_sphinx/config.py:178]: But shouldn't the extension know, at this point, whether the package that has this function was added or not?

File: docs/index.rst (hunk 5 above)
Comment (tigarmo) [reply docs/index.rst:15]: But isn't the canonical-sphinx documentation built with [the full set](https://github.com/canonical/canonical-sphinx/pull/50/files#diff-03efc769b870804394632e45d7885272b44c16939517fb31c9d7c614d2ffae57L27)?

File: canonical_sphinx/config.py (hunk 4 above)
Comment (SecondSkoll) [reply canonical_sphinx/config.py:178]: This feeds HTML context into the design override of Furo. The [HTML template override](https://github.com/canonical/canonical-sphinx/blob/856bd3388de23dcc99424f4c527b40fdde0a45ae/canonical_sphinx/theme/templates/footer.html#L75) has no real way of checking if the function exists - so if this is true by default it will call the function that doesnt exist.

The setup function is affected by the installed pip packages - and that establishes which extensions are loaded, but config_inited has no way of knowing - and for the most part thats fine, nothing will break if you set up an extension that isn't installed and isn't loaded. This seemed cleaner - and easier than passing some value to config_inited, or doing another check and conditionally changing things.

File: docs/index.rst (hunk 5 above)
Comment (SecondSkoll) [reply docs/index.rst:15]: I added a tox test for the minimal set which fails if this change isn't made.

We could add another set of docs to test against that includes all the functionality provided by the optional extensions - but a lot of the optional functions are tested downstream anyway.

File: canonical_sphinx/config.py (hunk 4 above)
Comment (tigarmo) [reply canonical_sphinx/config.py:178]: But then everybody that wants to display contributors will have to set this to `True`, which used to be the default, right? Isn't that undesirable? I thought the plan was to have this "selective loading" be more transparent to the user - that is, if they install `canonical-sphinx[full]` they get the full behavior with its defaults.

How about this: The extension that provides the `get_contributors_for_file()` function is `canonical.contributor-listing`, right? What if here in `config_inited()` you do

```python
has_contributor_listing = "canonical.contributor-listing" in app.extensions
...
html_context["has_contributor_listing"] = has_contributor_listing
```

... then update the template to read:

```
{% if has_contributor_listing and display_contributors and pagename and page_source_suffix %}
```

File: docs/index.rst (hunk 5 above)
Comment (tigarmo) [reply docs/index.rst:15]: alright, but moving forward I think we should be careful with changing these docs for the sake of testing. I know they're pretty bare-bones now but ideally they'll document the extension itself, so chances are they won't be build-able with the minimum set of dependencies.

But that can be for the future!

File: canonical_sphinx/config.py (hunk 4 above)
Comment (SecondSkoll) [reply canonical_sphinx/config.py]: I've made changes that implement your suggestion here, but I would note that:

* There are roughly as many private, internal documentation sets using the starter pack as public ones, and we do not always want to show contributors for these doc sets (so setting this to default True may not be desirable anyway - and we should actually put some effort into understanding why this default was chosen in the first place).
* The overall design of the contributor listing extension an is pretty poor - in that it relies on a specific template override in our design implementation and wouldn't be usable by other Sphinx users despite being a public extension. With plans to spin off the design overrides into a separate extension, this would mean that you'd need to be loading 3 separate extensions just to have that extension function.

File: docs/index.rst (hunk 5 above)
Comment (SecondSkoll) [reply docs/index.rst:15]: I don't think this extension will require a dedicated documentation site to provide information on the extension. It is not complex enough to require more than a detailed readme.

File: tox.ini (hunk 3 above)
Comment (SecondSkoll) [reply tox.ini]: Resolved with 5ab6bb5

File: canonical_sphinx/config.py (hunk 4 above)
Comment (tigarmo) [reply canonical_sphinx/config.py]: I'm fine with changing the defaults of whatever we wish (in my opinion, whatever the documentation team decides to do is fine with me), I'm just against changing a default here, on this unrelated PR. It was also an exercise in using selective dependencies, because I expect that there will possibly be dependency sets other than "minimal" and "full".

No comments on the usability of the contributor listing extension, whatever you decide to do with that is also fine with me :)

File: canonical_sphinx/config.py [hunk 7]
```diff
@@ -175,10 +175,14 @@ def config_inited(_app: Sphinx, config: Any) -> None:  # noqa: ANN401
         ("github_issues", "enabled"),
         ("discourse", "https://discourse.ubuntu.com"),
         ("sequential_nav", "none"),
-        ("display_contributors", False),
+        ("display_contributors", True),
     ]
 
+    has_contributor_listing = "canonical.contributor-listing" in _app.extensions
```
Comment (tigarmo) [inline canonical_sphinx/config.py:181]: now that we're using `_app` we should rename it to `app`

**Review by tigarmo — APPROVED** [review]
Thanks!

File: canonical_sphinx/config.py (hunk 4 above)
Comment (SecondSkoll) [reply canonical_sphinx/config.py]: The way the selecting loading works is that it will load anything on the list that's installed - I felt that was easier to implement than having more granularity in the optional dependencies.

So if you have sphinx-tabs alongside canonical-sphinx in your requirements file, it will be configured.

But you make a reasonable point about changing defaults on an unrelated PR, in retrospect I was just being lazy :P.

File: canonical_sphinx/config.py (hunk 7 above)
Comment (SecondSkoll) [reply canonical_sphinx/config.py:181]: I just hijacked what was already in place in the config_inited function (`def config_inited(_app: Sphinx, config: Any) -> None:  # noqa: ANN401`) I'm happy to change it but I'm not really across all the best practices.

I'm happy to change things if you have suggestions (and I'm always happy to learn), but you may have to be specific.

Comment (tigarmo) [reply canonical_sphinx/config.py:181]: Ah, I see. There's a linter that complains if a function/method parameter is not used. The convention (that the linter follows) is that parameters that aren't used on purpose are prefixed with underscore.
Now that the parameter is used, we can remove the underscore and the linter will be happy

Comment (SecondSkoll) [reply canonical_sphinx/config.py]: That makes sense. I've submitted a fix.
